aip/0008.md

```diff
@@ -174,6 +174,198 @@ When AIPs reference other AIPs, the prosaic text **must** use the format
+$.when($.ready)(() => {
+  const context = localStorage.get("context")
+
+  const anchors = window.getElementById("aip-main").getElementsByTagName("a")
+  let href
+  for(i = 0; i anchors.length; i++) {
+    href = anchors.getAttribute("href")
+
+    if( href is relative) {
+      if({{ file_exits("/"context"/"href) }})
+        use /context/href
+      else
+        use ./href
+    }
+  }
+})
+```
+
+Pros:
+
+- client does not have to do anything for context based references
+
+Cons:
+
+- client will need to do something different for static references
+- javascript iterates over all the anchors in the aip-main element, and
+  potentially alters them. This is just more overhead when loading a page. Not
+  huge, but still there.
+- feels a little hackish.
+
+#### Class Based
+
+Takes advantage of a kramdown feature for providing css classes to links
+inline.
+
+**What the AIP Editor Adds** Must be done when defining the reference:
+
+```
+[This][]{.context} is a context based reference.
+[This][]{.static} is a static reference.
+```
+
+**What the implementation looks like** implementation mostly looks like the
+automagical one, except we are looking for this class.
+
+```
+$.when($.ready)(() => {
+  const context = localStorage.get("context")
+
+  const anchors = window.getElementById("aip-main").getElementsByClass("context")
+  let href
+  for(i = 0; i anchors.length; i++) {
+    href = anchors.getAttribute("href")
+
+    if( href is relative) {
+      if({{ file_exits("/"context"/"href) }})
+        use /context/href
+      else
+        use ./href
+    }
+  }
+})
+```
+
+Pros:
+
+- declarative, similar solution for both static and context based references
+- optional, although the aip editor **should** use the declarative solution, if
+  they don't we can still treat the default as `static`.
+- feels less hackish as it takes advantage of existing functionality.
+
+Cons:
+
+- tighter coupling with kramdown.
+
+#### Failover Redirects
+
+Worth mentioning, but I have not gone too deep into it as I have the above
+solutions. We can also write our own `redirect.html` file in the `_layouts`
+directory and may be able to add some logic to fail over to a parent AIP if the
+context based one doesn't work.
+
+Pros:
+
+- Way to determine whether an AIP reference is context based or static may be
+  allowed to be more arbitrary and easier for aip editor to implement.
+
+Cons:
+
+- redirects in GitHub pages can only be done via meta-refresh. If a context
+  mostly inherits from it's parent except for one or two files, that means that
+  the request process for most context based AIP references would be as
+  follows:
+
+1. Request non existent page
+2. Jekyll redirects to our custom redirect page
+3. redirect page is returned to the browser, client should not see if it fall
+   goes well.
+4. redirect page creates new request for parent page.
+
+- The above process could be a pain if there are multiple parents (i.e. an
+  extension of an extension).
+
```

Comment:
Hopefully it's clear that not all of this would actually go into the AIP.

The Static AIP Reference and the Context Based Reference sections kind of describe the problem I am trying to solve, and the h4 sections below that describe some of the solutions I have considered. This is where I am looking for feedback.